pulser-core/pulser/backend/state.py

```diff
@@ -0,0 +1,236 @@
+        p_false_pos: float = 0.0,
+        p_false_neg: float = 0.0,
+    ) -> Counter[str]:
+        """Sample bitstrings from the state, taking into account error rates.
+
+        Args:
+            num_shots: How many bitstrings to sample.
+            one_state: The eigenstate that measures to 1. Can be left undefined
+                if the state's eigenstates form a known eigenbasis with a
+                defined "one state".
+            p_false_pos: The rate at which a 0 is read as a 1.
+            p_false_neg: The rate at which a 1 is read as a 0.
+
+        Returns:
+            The measured bitstrings, by count.
+        """
+        bitstring_probs = self.bitstring_probabilities(
+            one_state=one_state, cutoff=1 / (1000 * num_shots)
+        )
+        bitstrings = np.array(list(bitstring_probs))
+        probs = np.array(list(map(float, bitstring_probs.values())))
+        dist = np.random.multinomial(num_shots, probs)
+        # Filter out bitstrings without counts
+        non_zero_counts = dist > 0
+        bitstrings = bitstrings[non_zero_counts]
+        dist = dist[non_zero_counts]
+        if p_false_pos == 0.0 and p_false_neg == 0.0:
+            return Counter(dict(zip(bitstrings, dist)))
+
+        # Convert bitstrings to a 2D array
+        bitstr_arr = np.array([list(bs) for bs in bitstrings], dtype=int)
+        # If 1 is measured, flip_prob=p_false_neg else flip_prob=p_false_pos
+        flip_probs = np.where(bitstr_arr == 1, p_false_neg, p_false_pos)
+        # Repeat flip_probs of a bitstring as many times as it was measured
+        flip_probs_repeated = np.repeat(flip_probs, dist, axis=0)
+        # Generate random matrix of same shape
+        random_matrix = np.random.uniform(size=flip_probs_repeated.shape)
+        # Compare random matrix with flip probabilities to get the flips
+        flips = random_matrix < flip_probs_repeated
+        # Apply the flips with an XOR between original array and flips
+        new_bitstrings = bitstr_arr.repeat(dist, axis=0) ^ flips
+
+        # Count all the new_bitstrings
+        # Not converting to str right away because tuple indexing is faster
+        new_counts: Counter = Counter(map(tuple, new_bitstrings))
+        return Counter(
+            {"".join(map(str, k)): v for k, v in new_counts.items()}
+        )
+
+    @classmethod
+    @abstractmethod
+    def from_state_amplitudes(
+        cls: Type[StateType],
+        *,
+        eigenstates: Sequence[Eigenstate],
+        amplitudes: dict[str, ArgScalarType],
+    ) -> StateType:
+        """Construct the state from its basis states' amplitudes.
+
+        Args:
+            eigenstates: The basis states (e.g., ('r', 'g')).
+            amplitudes: A mapping between basis state combinations and
+                complex amplitudes.
+
+        Returns:
+            The state constructed from the amplitudes.
+        """
+        pass
+
+    def infer_one_state(self) -> Eigenstate:
```

Comment:
why do we not simply throw an error if the eigenbasis is not predefined? No backend would know how to simulate it anyway.